pydantic/types.py

```diff
@@ -687,6 +689,110 @@ def get_secret_value(self) -> bytes:
         return self._secret_value
 
 
+class PaymentCardBrand(Enum):
+    amex = 'American Express'
+    mastercard = 'Mastercard'
+    visa = 'Visa'
+    other = 'other'
+
+    def __str__(self) -> str:
+        return self.value
+
+
+class PaymentCardNumber(str):
+    """
+    Based on: https://en.wikipedia.org/wiki/Payment_card_number
+    """
+
+    strip_whitespace = True
+    min_length: int = 12
+    max_length: int = 19
+    bin: str
+    last4: str
+    brand: PaymentCardBrand
+
+    def __new__(cls, card_number: str) -> 'PaymentCardNumber':
+        obj = str.__new__(cls, card_number)  # type: ignore
+        obj.bin = card_number[:6]
+        obj.last4 = card_number[-4:]
+        obj.brand = cls.get_brand(card_number)
+        return obj
+
+    @classmethod
+    def __get_validators__(cls) -> 'CallableGenerator':
+        yield not_none_validator
+        yield str_validator
+        yield constr_strip_whitespace
+        yield constr_length_validator
+        yield cls.validate_digits
+        yield cls.validate_luhn_check_digit
+        yield cls.init
+        yield cls.validate_length_for_brand
+
+    @classmethod
+    def init(cls, card_number: str) -> 'PaymentCardNumber':
```

Comment:
I think you can remove this method and just have `yield cls` in the `__get_validators__`

Reply:
no luck ...

```
matin@herfei pydantic (payment_card) $ gd
diff --git a/pydantic/types.py b/pydantic/types.py
index 2928bc7..f13e92a 100644
--- a/pydantic/types.py
+++ b/pydantic/types.py
@@ -716,13 +716,9 @@ class PaymentCardNumber(str):
         yield constr_length_validator
         yield cls.validate_digits
         yield cls.validate_luhn_check_digit
-        yield cls.init
+        yield cls
         yield cls.validate_length_for_brand

-    @classmethod
-    def init(cls, card_number: str) -> 'PaymentCardNumber':
-        return cls(card_number)
-
     @property
     def bin(self) -> str:
         return self[:6]
matin@herfei pydantic (payment_card) $ pytest -s tests/test_types_payment_card_number.py
Test session starts (platform: darwin, Python 3.7.4, pytest 5.1.2, pytest-sugar 0.9.2)
rootdir: /Users/matin/code/pydantic, inifile: setup.cfg
plugins: sugar-0.9.2, cov-2.7.1, mock-1.10.4
collecting ...
――――――――――――――――――――――――――――――――――――――― tests/test_types_payment_card_number.py ――――――――――――――――――――――――――――――――――――――――
tests/test_types_payment_card_number.py:17: in <module>
    class PaymentCard(BaseModel):
pydantic/main.py:166: in __new__
    config=config,
pydantic/fields.py:153: in infer
    schema=schema,
pydantic/fields.py:122: in __init__
    self.prepare()
pydantic/fields.py:188: in prepare
    self._populate_validators()
pydantic/fields.py:285: in _populate_validators
    self.validators = self._prep_vals(v_funcs)
pydantic/fields.py:293: in _prep_vals
    return [make_generic_validator(f) for f in v_funcs if f]
pydantic/fields.py:293: in <listcomp>
    return [make_generic_validator(f) for f in v_funcs if f]
pydantic/class_validators.py:135: in make_generic_validator
    sig = signature(validator)
../../.pyenv/versions/3.7.4/lib/python3.7/inspect.py:3083: in signature
    return Signature.from_callable(obj, follow_wrapped=follow_wrapped)
../../.pyenv/versions/3.7.4/lib/python3.7/inspect.py:2833: in from_callable
    follow_wrapper_chains=follow_wrapped)
../../.pyenv/versions/3.7.4/lib/python3.7/inspect.py:2364: in _signature_from_callable
    'no signature found for builtin type {!r}'.format(obj))
E   ValueError: no signature found for builtin type <class 'pydantic.types.PaymentCardNumber'>

!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!! Interrupted: 1 errors during collection !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!

Results (0.26s):
```

Reply:
Ooh, that is interesting. I think adding `__init__` might fix it.

Reply:
forgot to rip out properties. fixed: d1397026d79b73f9bcb8e3f7ea8856cda57516b4